### ISSUE TITLE
Standalone Server

### ISSUE THREAD

**Issue by abetlen** [abetlen/llama-cpp-python#21]
Since the server is one of the goals / highlights of this project. I'm planning to move it into a subpackage e.g. `llama-cpp-python[server]` or something like that.

Work that needs to be done first:
- [x] Ensure compatibility with OpenAI
  - [x] Response objects match
  - [x] Request objects match
  - [x] Loaded model appears under `/v1/models` endpoint
  - [x] ~~Test OpenAI client libraries~~
  - [x] Unsupported parameters should be silently ignored
- [x] Ease-of-use
  - [x] Integrate server as a subpackage
  - [x] CLI tool to run the server
  
Future work
- [ ] Prompt caching to improve latency
- [ ] Support multiple models in the same server
- [ ] Add tokenization endpoints to make it easier to make it easier for small clients to calculate context window sizes

**Comment by MillionthOdin16** [comment]
Just a note, I found a package [fastapi-code-generator](https://koxudaxi.github.io/fastapi-code-generator) that you can put the OpenAI OpenAPI spec into and it will generate a server skeleton with the correct models and endpoints. Similarly, there are packages that can create test cases for the endpoints based off the API spec. This might save some time and we can return a not implemented error for endpoints that our server doesn't support.

**Comment by abetlen** [comment]
With the latest commit we now handle all the request parameters for the `/v1/completions`, `/v1/chat/completions` and `/v1/embeddings` endpoints. The server accepts additional parameters that are llama.cpp specific and ignores any that we currently don't support.

Next step is really just to bundle this into the PyPI package as a subpackage  so it can be installed with `pip install llama-cpp-python[server]` and then run with `python -m llama_cpp.server` or something like that.

**Comment by MillionthOdin16** [comment]
Awesome work! Just FYI, Llama CPP just got some major bug fixes that improves performance in the last hour. There should no longer be a performance degradation as the context size increases. Hopefully this translates into better performance for us too 🔥

**Comment by abetlen** [comment]
Awesome, I'll update the package!

**Comment by abetlen** [comment]
@MillionthOdin16 pushed the updated llama.cpp and the standalone server.

Do you mind testing it for me should?

Just update from pip and run `"MODEL=/path/to/model python3 -m llama_cpp.server"`

**Comment by MillionthOdin16** [comment]
@abetlen done! I created a #29 with some fixes, especially for windows. API is super nice. I did experience significantly slower `chat_completion` performance compared to the other endpoints (as you previously mentioned). But overall, super cool!

**Comment by MillionthOdin16** [comment]
One extra note on use, I think it would be nice to pass in the model (and eventually model folder) as an argument to llama_cpp.server instead of using an env var. Would make it more similar to other usages I think

**Comment by abetlen** [comment]
Next steps on the server (no particular order):
- [ ] multiple models selectable by model parameter
- [ ] prompt caching (if possible or maybe just hack this with multiple contexts)
- [ ] server cli options
- [ ] logprobs
- [ ] investigate adding `/models/{model}` endpoint
- [ ] model aliasing (kind of a hack but could fix some issues)

**Comment by riverzhou** [comment]
I want to save chat log.  
What's the best practices for this purpose?
